src/core/frames/frame_controller.ts

```diff
@@ -175,8 +175,12 @@ export class FrameController implements AppearanceObserverDelegate, FetchRequest
   }
 
   async requestSucceededWithResponse(request: FetchRequest, response: FetchResponse) {
-    await this.loadResponse(response)
-    this.resolveVisitPromise()
+    if (response.redirected && response.header("Turbo-Frame") == "_top") {
+      session.visit(response.location)
```

Comment:
Please correct me if I'm wrong, but from my understanding, fetch automatically follows all redirects and gives us the latest GET response. Once Turbo detects the `Turbo-Frame` header in the latest response, it will visit the same URL again. This leads to duplicate requests, and it breaks Rails' flash messages.

So, instead of revisiting the latest URL, we should replace the whole page with the provided content.

Reply:
That's a good point. In this case, we'd want to load the contents _and still_ to translate the `<turbo-frame>`-level navigation into a page-wide `Visit`. I think we want to call `session.visit(response.location, { response: { statusCode: response.statusCode, redirected: response.redirected, responseHTML: ... })` so that the `Visit` is in the session-level history, but does not result in an additional HTTP request.